support exporting attachments when specific post ids are requested

drzraf opened the pull request:
support exporting attachments when specific post ids are requested (assuming the post whose ids were specified were not attachment themselves)

`getenv()` to limit merge-conflict with other pending PR + fact that there may be long debates about how to pass the option, eg: `--attachments` / `--force-attachment` / `--post__in=+<num>,-<num>,...` / `--with-deps` / ...

drzraf commented:
About the change in `attachments_for_specific_post_types()`. This function should be made more generic (accept an array of filters?)
On overall, reusing `WP_Query()` may be a good choice to not duplicate query-builder logic?

schlessera commented:
Using `getenv()` is not a good choice here. All options for commands should be consistent and have proper command line arguments. Otherwise, we'll end up with countless custom constants over time.

In general, the process should be to open an issue first describing the problem that you're trying to solve, and then we can discuss details like the names of the arguments together, before you spend time writing code. See the section about the [pull request workflow in the handbook](https://make.wordpress.org/cli/handbook/pull-requests/).

I think going with a `--attachments` flag is pretty straightforward, provided it mimics the current behaviour.

So, the flag should default to `true`, _unless_ you specify specific post IDs with `--post__in`, in which case it should default to `false`.

This will not only allow what you have proposed above, but will also allow to _disable_ attachments for any of the other use case if needed.

What do you think?

danielbachhuber commented:
> I think going with a `--attachments` flag is pretty straightforward, provided it mimics the current behaviour.

One nit: `--attachments` is ambiguous. Something like `--with-attachments` is better, but still doesn't describe the fact that attachments are always included by default.

danielbachhuber commented on src/WP_Export_Query.php:
```diff
@@ -262,7 +266,7 @@ private function category_where() {
 
 	private function attachments_for_specific_post_types( $post_ids ) {
 		global $wpdb;
-		if ( !$this->filters['post_type'] ) {
```
In the WordPress admin, exporting a specific post type doesn't include attachments. We need to persist the current behavior here. If the `--with-attachments` flag is included, then we can include the attachments.

drzraf commented:
> > @@ -262,7 +266,7 @@ private function category_where() {
>  
>  	private function attachments_for_specific_post_types( $post_ids ) {
>  		global $wpdb;
> -		if ( !$this->filters['post_type'] ) {

I'd expect this hunk to cause trouble.
But I'd rather let you decide the future for this function and the way
this must be changed (if it really has to)

drzraf commented:
Thanks for the feedbacks about `--(with-)attachments`

PR #13 is 2 weeks old, all its issues have been addressed (except an inoffensive `wp_clear_object_cache()` that any maintainer may fix himself in order to merge (or fix later if it really needs to))
I'm really *not* willing to *maintain* many feature-branches and/or fighting to introduce *features* in a conflict-less  way with other pending patches (and even less willing to rebase)

For that reason the `getenv() to limit merge-conflict with other pending PR`
Although rerolls are sometimes really needed, many others about typo/details/... would really have have to be **non-merge-blocking** in order to ease fruitful contribution.
(and note the 10 minutes runs of the testsuite)

I'd really loved if you, knowledgeable maintainers, responsible of codebase quality, could manage the 5 or 10% left for a PR rather than expecting outside contributor to reroll the patch X times to reach a perfect the 100% quality expected. Would make wp-cli contribution experience much less frustrating for the best of the software.

For the context, I'm working on a plugin + continuous integration (to ensure code quality and avoid regression). wp-cli is just a *tool* for that end and I intend to make it more performant for the task in the hope it benefit to other WP dev, but I currently can't afford dozens of hours to wp-cli itself.

danielbachhuber commented:
> I'd really loved if you, knowledgeable maintainers, responsible of codebase quality, could manage the 5 or 10% left for a PR rather than expecting outside contributor to reroll the patch X times to reach a perfect the 100% quality expected. Would make wp-cli contribution experience much less frustrating for the best of the software.

All you need to do is ask :)

drzraf commented:
thx for the positive reply @danielbachhuber 
Added a commit with using `--with_attachments` and addressing reviews of @schlessera 
For time constraint and given that `wp post generate` does not offer a `--generate-attachments` may I let to you, maintainers, the work on the behat tests?

drzraf commented:
For the record I although wanted to include comments' images in my fixtures.
These images were the result of an ACF-field attached to comment (here inside the `"illustration"` ACF repeater field).

* The method is thus to make a first dump (`--with_comments`)
* Then running such an xpath expr:
`xpath -q -e '//*[wp:meta_key/text()[starts-with(. , "illustration_repeater_")]]/wp:meta_value/text()'`
* ... whose ID (ACF `meta_value` contains the attachment's `post_id`) could be added to a second `wp export --post__in=` run.

A long-term solution may be a hook allowing specific plugin to interpret the **semantic** of the `meta_value` for a specific `meta_key` (and, for example, add necessary post_id to the dump)

drzraf commented on src/Export_Command.php:
```diff
@@ -315,6 +324,18 @@ private function check_post__in( $post__in ) {
+		if ( is_null( $with ) )
+			return true;
+
+		if ( (int) $with <> 0 && (int) $with <> 1 ) {
```
yeah but currently, all option use a "_" between words, so that would mean --no-with_attachments

schlessera replied:
Hmm, good point. I think the underscore in this command came about because it is using some of the variable names from WP.

I'll discuss this in tomorrow's office hours to see what the best approach is.

schlessera replied:
I discussed this with @danielbachhuber , and we agreed to use `--with_attachments` as flag, knowing that the negation will be a weird `--no-with_attachments`.

schlessera replied:
Do you need assistance with converting the parameter into flag notation?

drzraf commented:
NB: I had no other option than fully recreate the branch because `master` changed too much since the original patch.

danielbachhuber commented:
My biggest concern with this PR is that we aren't introducing a change in the existing behavior. It's not quite clear to me based on a cursory review, so would be worthwhile to make sure there are tests around the current behavior.

schlessera commented:
I'll add tests by building the tests first, and only then applying the patch, just to stay on the safe side.

schlessera commented:
@danielbachhuber You can review and merge this PR now.

To make sure we don't change the default behaviour, I verified that the normal `post__in` export does not suddenly start adding attachments. This should be the only use case that could be impacted. That particular test succeeds both with or without this PR.

vadimkantorov commented:
Does `--with_attachments` have effect of actually embedding the attachments/image content inside the XML? (e.g. as CDATA/base64)

I'm interested in some portability between blogging platforms and the WXR/XML format could achieve this, if export/import can support embedding the images/attachments/resources inside the WXR.

swissspidy commented:
I don't think WordPress supports base64/cdata in WXR files. Therefore it's not something WP-CLI supports.

I see you just opened #118 so let's continue discussion there instead of a 7 year old closed PR :)
